### DISCUSSION TITLE
How to "Go to next symbol highlight"

### DISCUSSION THREAD

**Discussion by shybyte** [zed-industries/zed#21193] [Feature Requests]
When I move the cursor on a symbol in the editor, it highlights all occurrences of this symbol in the current file.
How can I move the cursor (jump) between these highlighted symbols?

![image](https://github.com/user-attachments/assets/7ee32af8-396b-4d1a-95f4-f682a308c1ab)

In VS-Code this functionality is called "Go to next symbol highlight" and it is very prominently mapped to F7.

Interestingly it's very easy  extend the (multi)selection to include the next occurrence:
 
![image](https://github.com/user-attachments/assets/53d7a1b9-619c-42ea-afea-88e0f1d12a10)

But I don't want to extend the selection, I just want to jump/navigate/toggle between all occurrences.

**Comment by garthdw** [comment]
It's not mapped by default, but you can do a custom keymap for `editor::SelectNext` but you also need to ensure the following option is set `{ "replace_newest": true }`.

e.g.
```json
[
	{
		"context": "Editor",
		"bindings": {
			"f7": ["editor::SelectNext", { "replace_newest": true }]
		}
	}
]
```

**Reply by MarcoPolo** [reply]
This isn't exactly the same. The original post should navigate between symbols highlighted by the LSP (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentHighlight)

**Comment by MarcoPolo** [comment]
I'll chime in here to explain my motivation for this feature. 

Having the ability to navigate directly to the next symbolHighlight/reference lets you follow the data flow of a variable. If you highlight the function itself (depending on the LSP), you can also navigate to all returns.

Note that this is a different feature from navigating to the next match, as that is not language-context aware. For example, if you have a bar named `foo` it would also navigate to an unrelated variable `fooBar`.

**Comment by MarcoPolo** [comment]
I have a patch the adds this feature. Here's how it works:

- The editor struct has a `background_highlights`.
- Collect all highlights with the keys `[DocumentHighlightRead, DocumentHighlightWrite]`
- Depending on the direction, move the cursor to the next or previous highlight relative to the current position.

**Reply by MarcoPolo** [reply]
https://github.com/zed-industries/zed/pull/35994

**Reply by shybyte** [reply]
@MarcoPolo Thank you!

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
